File: app/src/main/java/studio/gooduse/kitchenprep/data/SchedulingEngine.kt
package studio.gooduse.kitchenprep.data

import kotlin.math.max

class DependencyCycleException : IllegalStateException("Dependency cycle prevents automatic scheduling")

data class ScheduledTask(val taskId: String, val startAt: Long, val endAt: Long)

object SchedulingEngine {
    private data class Allocation(val start: Long, val end: Long, val units: Int)

    fun schedule(
        tasks: List<TaskEntity>,
        dependencies: List<TaskDependencyEntity>,
        taskResources: List<TaskResourceEntity>,
        resources: List<ResourceEntity>,
        timingMode: String?,
        targetReadyAt: Long?,
        now: Long,
    ): List<ScheduledTask> {
        if (tasks.isEmpty()) return emptyList()
        val taskById = tasks.associateBy { it.id }
        val deps = dependencies.filter { !it.overridden && it.taskId in taskById && it.dependencyTaskId in taskById }
        val order = topologicalOrder(tasks, deps)
        val criticalSeconds = criticalPathSeconds(order, deps, taskById)
        val desiredBase = if (timingMode == "SERVE_AT" || timingMode == "READY_BY") {
            (targetReadyAt ?: now) - criticalSeconds * 1000L
        } else now
        val base = max(now, desiredBase)

        val capacities = resources.associate { it.id to if (it.availability == "AVAILABLE") it.capacity else 0 }
        val reqByTask = taskResources.groupBy { it.taskId }
        val allocations = mutableMapOf<String, MutableList<Allocation>>()
        val result = linkedMapOf<String, ScheduledTask>()
        val depsByTask = deps.groupBy { it.taskId }

        for (task in order) {
            val depEnd = depsByTask[task.id].orEmpty().maxOfOrNull { d -> result[d.dependencyTaskId]?.endAt ?: base } ?: base
            val durationMs = max(0L, task.estimatedDurationSeconds ?: 0L) * 1000L
            var start = max(base, depEnd)
            start = findResourceSlot(start, durationMs, reqByTask[task.id].orEmpty(), capacities, allocations)
            val end = start + durationMs
            result[task.id] = ScheduledTask(task.id, start, end)
            reqByTask[task.id].orEmpty().forEach { r ->
                allocations.getOrPut(r.resourceId) { mutableListOf() }.add(Allocation(start, end, r.capacityUnits))
            }
        }

        if ((timingMode == "SERVE_AT" || timingMode == "READY_BY") && targetReadyAt != null) {
            val latestEnd = result.values.maxOfOrNull { it.endAt } ?: targetReadyAt
            val requestedShift = max(0L, latestEnd - targetReadyAt)
            val earliestStart = result.values.minOfOrNull { it.startAt } ?: now
            val safeShift = minOf(requestedShift, max(0L, earliestStart - now))
            if (safeShift > 0) return result.values.map { it.copy(startAt = it.startAt - safeShift, endAt = it.endAt - safeShift) }
        }
        return result.values.toList()
    }

    fun recomputeAvailability(tasks: List<TaskEntity>, dependencies: List<TaskDependencyEntity>): List<TaskEntity> {
        val terminal = tasks.associate { it.id to (it.status == "DONE" || it.status == "SKIPPED") }
        val depsByTask = dependencies.groupBy { it.taskId }
        return tasks.map { task ->
            if (task.status in setOf("ACTIVE", "WAITING", "DONE", "SKIPPED")) task else {
                val unresolved = depsByTask[task.id].orEmpty().any { d -> !d.overridden && terminal[d.dependencyTaskId] != true }
                task.copy(status = if (unresolved) "BLOCKED" else "AVAILABLE")
            }
        }
    }

    private fun topologicalOrder(tasks: List<TaskEntity>, dependencies: List<TaskDependencyEntity>): List<TaskEntity> {
        val byId = tasks.associateBy { it.id }
        val indegree = tasks.associate { it.id to 0 }.toMutableMap()
        val outgoing = mutableMapOf<String, MutableList<String>>()
        dependencies.forEach { d ->
            if (d.taskId == d.dependencyTaskId) throw DependencyCycleException()
            if (d.taskId in byId && d.dependencyTaskId in byId) {
                indegree[d.taskId] = indegree.getValue(d.taskId) + 1
                outgoing.getOrPut(d.dependencyTaskId) { mutableListOf() }.add(d.taskId)
            }
        }
        val ready = java.util.PriorityQueue<TaskEntity>(compareBy<TaskEntity> { it.sortOrder }.thenBy { it.id })
        tasks.filter { indegree[it.id] == 0 }.forEach(ready::add)
        val out = mutableListOf<TaskEntity>()
        while (ready.isNotEmpty()) {
            val task = ready.remove()
            out += task
            outgoing[task.id].orEmpty().forEach { child ->
                indegree[child] = indegree.getValue(child) - 1
                if (indegree[child] == 0) ready.add(byId.getValue(child))
            }
        }
        if (out.size != tasks.size) throw DependencyCycleException()
        return out
    }

    private fun criticalPathSeconds(order: List<TaskEntity>, deps: List<TaskDependencyEntity>, byId: Map<String, TaskEntity>): Long {
        val incoming = deps.groupBy { it.taskId }
        val longest = mutableMapOf<String, Long>()
        order.forEach { t ->
            val before = incoming[t.id].orEmpty().maxOfOrNull { longest[it.dependencyTaskId] ?: 0L } ?: 0L
            longest[t.id] = before + (byId[t.id]?.estimatedDurationSeconds ?: 0L)
        }
        return longest.values.maxOrNull() ?: 0L
    }

    private fun findResourceSlot(
        earliest: Long,
        durationMs: Long,
        requirements: List<TaskResourceEntity>,
        capacities: Map<String, Int>,
        allocations: Map<String, List<Allocation>>,
    ): Long {
        if (durationMs <= 0 || requirements.isEmpty()) return earliest
        var candidate = earliest
        var attempts = 0
        while (attempts < 1000) {
            attempts++
            var conflict = false
            for (req in requirements) {
                val cap = capacities[req.resourceId] ?: Int.MAX_VALUE
                val overlaps = allocations[req.resourceId].orEmpty()
                    .filter { it.start < candidate + durationMs && it.end > candidate }
                val used = overlaps.sumOf { it.units }
                if (used + req.capacityUnits > cap) {
                    conflict = true
                    val next = overlaps.minOfOrNull { it.end } ?: (candidate + durationMs)
                    candidate = if (next > candidate) next else candidate + 1L
                    break
                }
            }
            if (!conflict) return candidate
        }
        // Safety fallback: preserve liveness if resource contention is pathological.
        return earliest
    }
}

File: flutter_app/lib/domain/scheduling_engine.dart
import 'dart:math' as math;

import 'kitchen_models.dart';

class DependencyCycleException implements Exception {
  const DependencyCycleException();
  @override
  String toString() => 'Dependency cycle prevents automatic scheduling';
}

class ResourceUnavailableException implements Exception {
  const ResourceUnavailableException(this.resourceId);
  final String resourceId;
  @override
  String toString() => 'Resource unavailable: $resourceId';
}

class ScheduledTask {
  const ScheduledTask({
    required this.taskId,
    required this.startAtEpochMs,
    required this.endAtEpochMs,
  });
  final String taskId;
  final int startAtEpochMs;
  final int endAtEpochMs;
}

class ScheduleResult {
  const ScheduleResult({
    required this.tasks,
    required this.earliestFinishEpochMs,
    required this.latenessMs,
  });
  final List<ScheduledTask> tasks;
  final int earliestFinishEpochMs;
  final int latenessMs;

  bool get isLate => latenessMs > 0;
}

class _Allocation {
  const _Allocation(this.start, this.end, this.units);
  final int start;
  final int end;
  final int units;
}

class SchedulingEngine {
  const SchedulingEngine._();

  static ScheduleResult schedule(KitchenBoard board, int nowEpochMs) {
    if (board.tasks.isEmpty) {
      return ScheduleResult(
        tasks: const <ScheduledTask>[],
        earliestFinishEpochMs: nowEpochMs,
        latenessMs: 0,
      );
    }
    final ordered = _topologicalOrder(board.tasks);
    final criticalMs = _criticalPathMs(ordered, board);
    final target = board.timingMode == TimingMode.readyAt
        ? board.targetReadyAtEpochMs
        : null;
    final desiredBase = target == null ? nowEpochMs : target - criticalMs;
    final base = math.max(nowEpochMs, desiredBase).toInt();

    final allocations = <String, List<_Allocation>>{};
    final result = <String, ScheduledTask>{};

    for (final task in ordered) {
      var depEnd = base;
      for (final dependencyId in task.dependencyIds) {
        final dependency = result[dependencyId];
        if (dependency != null) depEnd = math.max(depEnd, dependency.endAtEpochMs).toInt();
      }
      final durationMs = math.max(0, task.durationSeconds ?? 0).toInt() * 1000;
      var start = math.max(base, depEnd).toInt();
      start = _findResourceSlot(board, task, start, durationMs, allocations);
      final end = start + durationMs;
      result[task.id] = ScheduledTask(
        taskId: task.id,
        startAtEpochMs: start,
        endAtEpochMs: end,
      );
      for (final requirement in task.resourceRequirements) {
        allocations
            .putIfAbsent(requirement.resourceId, () => <_Allocation>[])
            .add(_Allocation(start, end, requirement.units));
      }
    }

    final values = <ScheduledTask>[for (final task in ordered) result[task.id]!];
    final finish = values.map((item) => item.endAtEpochMs).fold<int>(
      nowEpochMs,
      (value, end) => math.max(value, end).toInt(),
    );
    final lateness = target == null ? 0 : math.max(0, finish - target).toInt();
    return ScheduleResult(
      tasks: values,
      earliestFinishEpochMs: finish,
      latenessMs: lateness,
    );
  }

  static void applySuggestions(KitchenBoard board, int nowEpochMs) {
    final result = schedule(board, nowEpochMs);
    final byId = {for (final item in result.tasks) item.taskId: item};
    for (final task in board.tasks) {
      final scheduled = byId[task.id];
      task.suggestedStartEpochMs = scheduled?.startAtEpochMs;
      task.suggestedEndEpochMs = scheduled?.endAtEpochMs;
    }
  }

  static void recomputeAvailability(KitchenBoard board) {
    final terminal = <String, bool>{
      for (final task in board.tasks) task.id: task.isTerminal,
    };
    for (final task in board.tasks) {
      if (task.status == TaskStatus.running || task.isTerminal) continue;
      final unresolved = task.dependencyIds.any((id) => terminal[id] != true);
      task.status = unresolved ? TaskStatus.blocked : TaskStatus.ready;
    }
  }

  static List<KitchenTask> _topologicalOrder(List<KitchenTask> tasks) {
    final byId = {for (final task in tasks) task.id: task};
    final indegree = {for (final task in tasks) task.id: 0};
    final outgoing = <String, List<String>>{};
    for (final task in tasks) {
      for (final dependencyId in task.dependencyIds) {
        if (!byId.containsKey(dependencyId)) continue;
        if (dependencyId == task.id) throw const DependencyCycleException();
        indegree[task.id] = (indegree[task.id] ?? 0) + 1;
        outgoing.putIfAbsent(dependencyId, () => <String>[]).add(task.id);
      }
    }
    final ready = <KitchenTask>[
      for (final task in tasks)
        if ((indegree[task.id] ?? 0) == 0) task,
    ];
    ready.sort((a, b) => _taskOrder(tasks, a, b));
    final output = <KitchenTask>[];
    while (ready.isNotEmpty) {
      final task = ready.removeAt(0);
      output.add(task);
      for (final child in outgoing[task.id] ?? const <String>[]) {
        indegree[child] = (indegree[child] ?? 0) - 1;
        if (indegree[child] == 0) {
          ready.add(byId[child]!);
          ready.sort((a, b) => _taskOrder(tasks, a, b));
        }
      }
    }
    if (output.length != tasks.length) throw const DependencyCycleException();
    return output;
  }

  static int _taskOrder(List<KitchenTask> tasks, KitchenTask a, KitchenTask b) {
    final priority = b.priority.index.compareTo(a.priority.index);
    if (priority != 0) return priority;
    return tasks.indexOf(a).compareTo(tasks.indexOf(b));
  }

  static int _criticalPathMs(List<KitchenTask> order, KitchenBoard board) {
    final longest = <String, int>{};
    for (final task in order) {
      var before = 0;
      for (final dependencyId in task.dependencyIds) {
        before = math.max(before, longest[dependencyId] ?? 0).toInt();
      }
      longest[task.id] =
          before + math.max(0, task.durationSeconds ?? 0).toInt() * 1000;
    }
    return longest.values.fold<int>(
      0,
      (value, duration) => math.max(value, duration).toInt(),
    );
  }

  static int _findResourceSlot(
    KitchenBoard board,
    KitchenTask task,
    int earliest,
    int durationMs,
    Map<String, List<_Allocation>> allocations,
  ) {
    if (durationMs <= 0 || task.resourceRequirements.isEmpty) return earliest;
    for (final requirement in task.resourceRequirements) {
      final resource = board.resourceById(requirement.resourceId);
      if (resource == null || !resource.available || requirement.units > resource.capacity) {
        throw ResourceUnavailableException(requirement.resourceId);
      }
    }
    var candidate = earliest;
    for (var attempts = 0; attempts < 10000; attempts++) {
      int? nextCandidate;
      for (final requirement in task.resourceRequirements) {
        final resource = board.resourceById(requirement.resourceId)!;
        final overlaps = (allocations[requirement.resourceId] ?? const <_Allocation>[])
            .where((item) => item.start < candidate + durationMs && item.end > candidate)
            .toList();
        final used = overlaps.fold<int>(0, (sum, item) => sum + item.units);
        if (used + requirement.units > resource.capacity) {
          final next = overlaps.map((item) => item.end).fold<int?>(null, (value, end) {
            if (value == null || end < value) return end;
            return value;
          });
          nextCandidate =
              math.max(candidate + 1, next ?? candidate + durationMs).toInt();
          break;
        }
      }
      if (nextCandidate == null) return candidate;
      candidate = nextCandidate;
    }
    throw StateError('Unable to find a resource-capacity-safe schedule.');
  }
}
